Product Collection - New 'No Results' block with default UI

imanish003 opened the pull request:
<!-- Please do not remove any information from this pull request. Instead, add N/A or leave blank if not applicable -->

## What

This PR introduces a custom 'No Results' block for the Product Collection block, specifically designed to replace the default core 'query-no-results' block usage. This new block provides the following default view when there are no products for selected filters.

![image](https://github.com/woocommerce/woocommerce-blocks/assets/16707866/e618cce6-9fd7-4b40-bc40-0e342cc92799)


Part of #10826

## Why

Developing a new "No Results" block is essential for an improved default user interface. This initiative will lay the groundwork for enhancing the user experience for merchants. Additionally, it will facilitate the creation of extensibility options for third-party developers (3PDs). Furthermore, we plan to introduce a pattern selection feature, enabling merchants to customize the appearance of the "No Results" view. For more info, you may also check Github Discussion #10707.

<!-- Describe the reason for your changes. This will help the reviewer and future readers get additional context -->

## Testing Instructions

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

_Please consider any edge cases this change may have, and also other areas of the product this may impact._

### Case 1: Normal

1. Create a new post & add Product Collection block
2. Verify that there is "No results" block, which UI should look like this:
![image](https://github.com/woocommerce/woocommerce-blocks/assets/16707866/6a52b7d6-1323-44f5-a71c-6dee5a6e0a97)

    
3. Save the post & verify on Frontend that "No results" blocks work as expected:
    - **Note:** This block will be only visible if there are no products to show on the frontend. You can achieve this by changing the filters in Inspector control such that there are no products to show for those filters. For example, try setting "Stock Status" filter to only "On backorder".
4. There are two links on frontend:
    - **store's home:** Verify that "store's home" link takes you to the home page of the store.
    - **clearing any filters:** This functionality is tested in Case 2.

### Case 2: Testing `clearing any filters` link

1. Create a new post.
2. Add filters blocks, i.e. Filter by Price, Filter by Stock, Filter by Attribute, Filter by Rating, etc. You can use a pattern called "Product Filters" which contains all the filters.
3. Add Product Collection block & publish the post
4. Go to frontend:
    - Change all filter block values so that there are no products for selected filters and no results block becomes visible. URL will look something like this: `http://store.local/162-2/?filter_stock_status=instock&filter_color=blue%2Cgray&query_type_color=or&min_price=29&max_price=32`
    - Click on "clearing any filters" link & verify that it clears all filter values.
    - *TO BE TESTED BY DEVELOPERS*
        - You can also add some random query variables to URL, for example `&random=value` & URL will look something like this: `http://store.local/162-2/?filter_stock_status=instock&filter_color=blue%2Cgray&query_type_color=or&min_price=29&max_price=32&random=value`
            - Verify that clicking on `clearing any filters` only removes query variables that are related to filters & doesn't remove this random query variable we added.

### Case 3: Adding styles to "No results" block

1. Create a new post & add Product Collection block.
2. Select "No results" block & change styles using Inspector control, i.e.:
    - Text, Background & Link color
    - Typography: Font family, appearance, line height, decoration etc.
3. Verify the changes made to these styles are visible on Editor.
4. Publish the post & verify that changes are visible on Frontend as well.

### Case 4: Adding/removing blocks

1. Create a new post & add Product Collection block.
2. In “No results” block:
    1. Verify that you can remove existing blocks.
    2. Verify that you can add new blocks to it.
    3. Verify that the above changes are reflected correctly on the frontend.

Here is a quick demo:


https://github.com/woocommerce/woocommerce-blocks/assets/16707866/a5907a6f-8131-45f9-9c18-6f693cd02b5c




* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested (ie: it makes changes to tests, coding standards, docblocks, etc.). -->
* [ ] Should be tested by the development team exclusively <!-- Check this box if this PR should be tested by the development team exclusively (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

## Screenshots or screencast

<!-- Any screenshots of UI changes will be helpful to include here. Leave blank if not applicable. -->

| Before | After |
| ------ | ----- |
|        |       |

## WooCommerce Visibility

<!-- Check this documentation link (../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WooCommerce core, part of the feature plugin, or experimental. -->
Required:

* [X] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental
* [ ] N/A

## Checklist

Required:

* [X] This PR has either a `[type]` label or a `[skip-changelog]` label.
* [ ] This PR is assigned to a milestone.

Conditional:

* [x] This PR has a UI change and has been cross-browser tested at different viewport sizes on both the frontend and in the editor.
* [x] This PR has a changelog description (if `[skip-changelog]` label is not present).
* [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR adds/removes an experimental interfaces, and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR has been accessibility tested.
* [ ] This PR has had any necessary documentation added/updated.

## Changelog
<!-- Provide a brief, descriptive summary of the changes in this PR. Include potential impacts on different parts of the product. Example: "Updated the checkout process to streamline the experience for users and reduce the number of steps." -->

> Product Collection - New 'No Results' block with default UI

imanish003 committed:
Add custom 'No Results' block for Product Collection

This commit introduces a new 'No Results' block, replacing the default core 'query-no-results' block within product collection block. The 'No Results' block provides a tailored experience for WooCommerce stores, displaying a custom message when no products are found in a query.

Key changes include:
- A new block type 'woocommerce/no-results' is registered with a complete configuration and content structure.
- Edit and save functions are implemented for the block, allowing for custom content and styles within the block editor and on the front end.
- Webpack entries are updated to include the new block in the build process.
- A PHP class NoResults is added for server-side rendering, which only displays the block when the product query returns no results, enhancing performance.
- The ProductCollectionUtils utility class is extended to support the new block's query needs.

The new block enriches the user experience by providing clearer communication when no products match the collection criteria and allows store owners to customize the message and presentation.

imanish003 committed:
Merge branch 'trunk' into update/10826-product-collection-enhance-no-results-view

imanish003 committed:
Update description

kmanijak commented on src/BlockTypes/NoResults.php:
```diff
@@ -0,0 +1,141 @@
+	 *
+	 * @var string
+	 */
+	protected $block_name = 'no-results';
```
May be worth adding a prefix suggesting this block belongs to Product Collection. What do you think?

```suggestion
	protected $block_name = 'product-collection-no-results';
```

imanish003 replied:
Agreed! Made the changes in commit https://github.com/woocommerce/woocommerce-blocks/pull/11783/commits/9dc01a9b4adce5cde28645de37bc4d791bd8cbcd.

kmanijak commented on bin/webpack-entries.js:
```diff
@@ -50,6 +50,9 @@ const blocks = {
 	'product-category': {},
 	'product-categories': {},
 	'product-collection': {},
+	'no-results': {
```
May be worth adding a prefix suggesting this block belongs to Product Collection. What do you think?

```suggestion
	'product-collection/no-results': {
```

imanish003 replied:
Agreed! Made the changes in commit https://github.com/woocommerce/woocommerce-blocks/pull/11783/commits/9dc01a9b4adce5cde28645de37bc4d791bd8cbcd.

kmanijak commented on src/BlockTypesController.php:
```diff
@@ -239,6 +239,7 @@ protected function get_block_types() {
 			'ProductCategories',
 			'ProductCategory',
 			'ProductCollection',
+			'NoResults',
```
May be worth adding a prefix suggesting this block belongs to Product Collection. What do you think?

```suggestion
			'ProductCollectionNoResults',
```

imanish003 replied:
Made the changes in [0a4395a](https://github.com/woocommerce/woocommerce-blocks/pull/11783/commits/0a4395a839aaddc4d4cdfdc0bf3e656f8244f53e)

kmanijak reviewed (approved):
I tested the changes and it works nice! Great work!

I'm thinking about prefixing "No Results" block name in couple of places to indicate it's only created for Product Collection block.

Pre-approving anyway!

imanish003 committed:
Merge branch 'trunk' into update/10826-product-collection-enhance-no-results-view

imanish003 committed:
Rename 'no-results' block to 'product-collection-no-results'

Changes include:
- Updating the block name in `edit.tsx` within the `product-collection` directory.
- Modifying the block.json file in the `no-results` inner block to reflect the new name.
- Adjusting the className in `edit.tsx` for the `no-results` inner block.
- Altering the webpack entries in `webpack-entries.js` to recognize the new block name.
- Renaming the block in `NoResults.php` to align with this update.

This renaming aims to make the block's purpose more clear and to maintain a consistent naming scheme within our product collection blocks.

imanish003 committed:
Rename NoResults to ProductCollectionNoResults for clarity

This commit renames the `NoResults` class to `ProductCollectionNoResults`. The change aims to enhance clarity and specificity about the class's purpose, indicating that it specifically handles no-result scenarios within product collections.

Changes made:
- Renamed `NoResults.php` to `ProductCollectionNoResults.php`.
- Updated the class name from `NoResults` to `ProductCollectionNoResults` in the file.
- Modified the reference in `BlockTypesController.php` to use the new class name.

This renaming ensures better readability and understanding of the class's role in the context of product collections. The primary change is the renaming, with no significant alterations in the class functionality.

imanish003 commented:
Hi @kmanijak, I've implemented the change you requested. Could you please review the PR once more before I proceed with merging it into trunk? Thank you 🙌🏻

xristos3490 reviewed (commented):
Nice work, @imanish003 👌 

I have two comments based on the screenshot below:
![NEddpH.png](https://github.com/woocommerce/woocommerce-blocks/assets/1105590/7188e53a-9986-4af1-9e1a-e1ecec3e1955)

1. How about removing the dot from the heading?
2. These inline paragraphs are creating a strange gap between words. Would it be possible to concat everything into a single paragraph block?

imanish003 committed:
Update No-Results Message Formatting in Product Collection Block

This commit simplifies the layout and message content for the 'No results found' message in the product collection block's no-results edit component. The changes include:

1. Removal of the full stop in the 'No results found' string for consistency.
2. Replacing the 'core/group' block with a 'core/paragraph' block.
3. Streamlining the message content to be more concise and integrated into fewer text blocks.
4. Direct links for 'clearing any filters' and navigating to the 'store's home' are now included in the same paragraph.

imanish003 commented:
Great spot, @xristos3490 – thank you! 🙏🏻
I've implemented the changes in commit a606add. Could you please review it once more?

xristos3490 reviewed (approved):
Looking good! Thanks, @imanish003!

kmanijak commented:
> Hi @kmanijak, I've implemented the change you requested. Could you please review the PR once more before I proceed with merging it into trunk? Thank you 🙌🏻

Rechecked after recent changes and all looks good!

PS. Great catches @xristos3490! 🙌

imanish003 committed:
Merge branch 'trunk' into update/10826-product-collection-enhance-no-results-view

wavvves committed:
Merge branch 'trunk' into update/10826-product-collection-enhance-no-results-view

wavvves committed:
Merge branch 'trunk' into update/10826-product-collection-enhance-no-results-view
